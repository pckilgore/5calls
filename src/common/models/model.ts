import { Issue, Category } from './issue';

export interface Outcome {
  label: string;
  status: string;
}

export interface VoterContact {
  id: string;
  name: string;
  location: string;
  phone: string;
}

export class CategoryMap {
  category: Category;
  issues: Issue[];
}

export type Party = 'democrat' | 'republican' | 'independent' | '';

export interface FieldOffice {
  city: string;
  phone: string;
}

export interface UserStat {
  all: string[];
  contactedCount: number;
  voiceMailCount: number;
  unavailableCount: number;
}

/**
 * Represents the place used to get location.
 * It may be one of three options:
 * 1. CACHED_ADDRESS - address stored in local storage
 * 2. BROWSER_GEOLOCATION - address obtained from the browser geolocation API
 * 3. IP_INFO - address obtained from ipinfo.io API
 */
export enum LocationFetchType {
  CACHED_ADDRESS = 'CACHED_ADDRESS', // 'address' in Choo version
  BROWSER_GEOLOCATION = 'BROWSER_GEOLOCATION', // 'browserGeolocation' in Choo version
  IP_INFO = 'IP_INFO' // 'ipAddress' in Choo version
}

/**
 * Encapsulates a location using latitude
 * and longitude. Undefined for either longitude
 * or latitude indicates that the geolocation has
 * has not been set.
 */
export interface GeolocationPosition {
  longitude: number | undefined;
  latitude: number | undefined;
}

/* 5 Calls API data */
export interface IssueData {
  issues: Issue[];
}

export interface GroupIssues {
  splitDistrict: boolean;
  invalidAddress: boolean;
  normalizedLocation: string | undefined;
  divisions: string[];
  issues: Issue[];
}

export interface CountData {
  count: number; // total call count
}

/* Data from iponfo.io API */
export interface IpInfoData {
  ip: string;
  hostname: string;
  city: string;
  region: string; // state
  country: string;
  loc: string; // long, lat - used in issue lookup
  org: string; // internet service provider
  postal: string; // zip code
}

export interface DonationGoal {
  goal: Donations; //
}

export interface Donations {
  count: number; // number of donors
  amount: number; // total collected (api===amount)
  total: number; // goal (api===total)
  kind: string; // denomincation (dollars)
}

export interface OutcomeButton {
  title: string;
  emoji: string;
  key: string;
}
