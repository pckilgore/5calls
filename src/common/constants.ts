export const APP_NAME = '5 Calls';
export const APP_URL = 'https://5calls.org';
// export const APP_URL = 'http://localhost:8090';
export const API_URL = 'https://api.5calls.org/v1';
// export const API_URL = 'http://localhost:8090/v1';
export const ISSUES_API_URL = `${API_URL}/issues`;
export const REPS_API_URL = `${API_URL}/reps`;
export const REPORT_API_URL = `${APP_URL}/report`;
export const STATS_API_URL = `${API_URL}/users/stats`;
export const PROFILE_API_URL = `${API_URL}/users/profile`;
export const DONATIONS_API_URL = `${API_URL}/donations`;
export const MIDTERMS_API_URL = `${API_URL}/midterms`;
export const GROUP_API_URL = `${API_URL}/groups`;
export const CONTACTS_API_URL = `${API_URL}/contacts`;
export const REMINDER_API_URL = `${API_URL}/remind`;
export const DONATE_URL = 'https://secure.actblue.com/donate/5calls-donate';
export const WRITER_APPLY_URL = 'https://airtable.com/shrJBbpixnJ09iVb3';
export const CODE_PROJECT_URL =
  'https://github.com/5calls/5calls/wiki/Getting-Involved-with-5-Calls-Development';

export const HELP_SET_LOCATION =
  'https://intercom.help/5calls/using-the-5-calls-website-and-apps/how-do-i-set-my-location-on-5-calls';
export const HELP_OTHER_REPRESENTATIVES =
  'https://intercom.help/5calls/using-the-5-calls-website-and-apps/why-do-i-only-see-some-of-my-representatives';

export const SHARE_BUCKET_URL =
  'http://5callsorg-shareimages.s3-website-us-west-1.amazonaws.com/';

const callbackURI = () => {
  if (window.location.host.includes('localhost')) {
    return 'http://localhost:3000/auth0callback';
  } else if (window.location.host.includes('test.5calls.org')) {
    return 'https://test.5calls.org/auth0callback';
  }
  return 'https://5calls.org/auth0callback';
};

export const Auth0Config = {
  domain: '5callsos.auth0.com',
  clientId: 'bVxfNRGD4azuJLWPDPHNNsfMElENPKjF',
  audience: 'https://5callsos.auth0.com/userinfo',
  callbackUri: callbackURI(),
  poweredURL: 'https://5calls.org',
  popupAuth: false
};

export const zipCodeRegex: RegExp = /^\d{5}(-\d{4})?$/;

export const cacheTimeout = {
  default: 24 * 60 * 60 * 1000, // 1 day
  groups: 24 * 60 * 60 * 1000 // 1 day
};

export const contact = {
  email: 'make5calls@gmail.com',
  github: 'https://github.com/5calls',
  twitter: 'https://twitter.com/make5calls',
  facebook: 'https://www.facebook.com/make5calls',
  apps: 'https://crgj.app.link/7R2bEB0R4F'
};
