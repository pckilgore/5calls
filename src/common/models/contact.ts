import { Party, FieldOffice } from './model';

export class Contact {
  id: string;
  name: string;
  phone: string;
  photoURL?: string;
  party: Party;
  state: string;
  reason: string;
  area?: string;
  // tslint:disable-next-line:variable-name
  field_offices?: FieldOffice[];

  public contactDisplay(): string {
    if (this.partyStateAbbr() !== '') {
      return this.name + ' (' + this.partyStateAbbr() + ')';
    }

    return this.name;
  }

  public partyStateAbbr(): string {
    if (this.party && this.state !== '') {
      const partyLetter = this.party.substring(0, 1);
      return partyLetter.toUpperCase() + '-' + this.state.toUpperCase();
    }

    return '';
  }
}

export const mockContact = Object.assign(new Contact(), {
  id: 123,
  name: 'Mock Contact III',
  phone: '4155551212',
  photoURL: undefined,
  party: 'Democrat',
  state: 'CA',
  reason: 'This is your mock representative',
  area: undefined,
  field_offices: []
});
