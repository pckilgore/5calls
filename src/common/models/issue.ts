import { Outcome, Contact, ContactList } from './';

export class Issue {
  public id: number;
  public name: string;
  public contactType: string;
  public contactAreas: string[];
  public reason: string;
  public script: string;
  public categories: Category[];
  public active: boolean;
  public createdAt: string;
  public slug: string;
  public outcomeModels: Outcome[];
  public link: string;
  public linkTitle: string;
  public stats: IssueStats;

  public slugOrID(): string {
    if (this.slug !== '') {
      return this.slug;
    }

    return this.id.toString();
  }

  public numberOfContacts(contactList: ContactList): number {
    return this.filteredContacts(contactList).length;
  }

  public currentContact(
    contactList: ContactList,
    contactIndex: number
  ): Contact | undefined {
    const contacts = this.filteredContacts(contactList);

    if (contactIndex <= contacts.length) {
      return contacts[contactIndex];
    }

    return undefined;
  }

  // why do we need filtered contacts if we already have an array of contacts?
  // a) ensuring consistent ordering
  // b) sometimes we get back multiple house reps in split areas
  public filteredContacts(contactList: ContactList): Contact[] {
    const contacts: Contact[] = [];

    for (const contactArea of this.contactAreas) {
      if (contactArea === 'US Senate') {
        contacts.push(...contactList.senateReps());
      } else if (contactArea === 'US House') {
        contacts.push(...contactList.houseRep());
      } else if (contactArea === 'Governor') {
        contacts.push(...contactList.governor());
      } else if (contactArea === 'StateUpper') {
        contacts.push(...contactList.stateUpper());
      } else if (contactArea === 'StateLower') {
        contacts.push(...contactList.stateLower());
      } else if (contactArea === 'SecretaryOfState') {
        contacts.push(...contactList.secState());
      } else if (contactArea === 'AttorneyGeneral') {
        contacts.push(...contactList.attyGeneral());
      }
    }

    return contacts;
  }

  public listItemLabel(): string {
    switch (this.contactType) {
      case 'REPS':
        return 'Call reps' + this.countLabel();
      case 'ACTION':
        return 'Take action' + this.countLabel();
      default:
        return '???';
    }
  }

  public countLabel(): string {
    if (this.stats.calls < 10) {
      return '';
    }

    switch (this.contactType) {
      case 'REPS':
        return ' ⋆ ' + this.stats.calls + ' calls made';
      case 'ACTION':
        return ' ⋆ ' + this.stats.calls + ' actions taken';
      default:
        return '???';
    }
  }
}

export interface IssueStats {
  completion: number;
  calls: number;
}

export interface Category {
  name: string;
}

export const mockIssue = Object.assign(new Issue(), {
  id: 123,
  name: 'Issue Name',
  contactType: 'REPS',
  contactAreas: [],
  reason: 'reason',
  script: 'issue *mock script*',
  active: true,
  createdAt: '',
  slug: 'issue-name',
  outcomeModels: [],
  link: '',
  linkTitle: '',
  stats: { completion: 0, calls: 10 }
});
