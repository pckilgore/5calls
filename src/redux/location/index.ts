export {
  LocationAction,
  LocationActionType,
  LocationClearedAction,
  LocationSetAction,
  SetLocationFetchTypeAction,
  SetSplitDistrictAction,
  CacheCityAction,
  NewLocationLookupAction
} from './action';
export {
  setLocation,
  clearAddress,
  setCachedCity,
  setLocationFetchType,
  setSplitDistrict
} from './actionCreator';
export { LocationState, locationStateReducer } from './reducer';
