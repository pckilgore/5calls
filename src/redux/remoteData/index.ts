export {
  IssuesAction,
  RemoteDataAction,
  RemoteDataActionType,
  CallCountAction
} from './action';
export { issuesActionCreator, callCountActionCreator } from './actionCreator';
export { RemoteDataState, remoteDataReducer } from './reducer';
export {
  startup,
  getIssuesIfNeeded,
  fetchCallCount
} from './asyncActionCreator';
