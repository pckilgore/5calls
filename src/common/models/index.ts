import { Issue, Category } from './issue';
import { Contact } from './contact';
import { ContactList } from './contactList';
import {
  Outcome,
  OutcomeButton,
  LocationFetchType,
  FieldOffice,
  CategoryMap,
  GeolocationPosition,
  CountData
} from './model';

export {
  Issue,
  Category,
  Contact,
  ContactList,
  Outcome,
  OutcomeButton,
  LocationFetchType,
  FieldOffice,
  CategoryMap,
  GeolocationPosition,
  CountData
};
