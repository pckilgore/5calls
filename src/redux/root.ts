import storage from 'redux-persist/lib/storage';
import getStoredState from 'redux-persist/lib/integration/getStoredStateMigrateV4';

import { ApplicationState } from './root';
import { LocationState, locationStateReducer } from './location';
import { CallState, callStateReducer } from './callState';
import { RemoteDataState, remoteDataReducer } from './remoteData';
import { UserStatsState, userStatsReducer } from './userStats';
import { UserState, userStateReducer } from './userState';
import { persistCombineReducers } from 'redux-persist';
import { defaultRemoteDataState } from './remoteData/reducer';
import { defaultUserStatsState } from './userStats/reducer';

export enum OutcomeType {
  UNAVAILABLE = 'unavailable',
  VOICEMAIL = 'voicemail',
  CONTACT = 'contact',
  SKIP = 'skip'
}

export interface ApplicationState {
  remoteDataState: RemoteDataState;
  callState: CallState;
  locationState: LocationState;
  userStatsState: UserStatsState;
  userState: UserState;
}

export const DefaultApplicationState: ApplicationState = {
  remoteDataState: defaultRemoteDataState,
  callState: {} as CallState,
  locationState: {} as LocationState,
  userStatsState: defaultUserStatsState,
  userState: {} as UserState
};

// DANGER: TypeScript magic ahead!!
// These articles should help your understanding of this section:
// https://blog.mariusschulz.com/2017/01/06/typescript-2-1-keyof-and-lookup-types
// https://blog.mariusschulz.com/2017/01/20/typescript-2-1-mapped-types
// https://moin.world/2017/06/18/10-TypeScript-Features-You-Might-Not-Know/
//
// This will be used by redux-persist in store.ts
// A type that is an ApplicationState key
export type ApplicationStateKeyType = keyof ApplicationState;
// A TS mapped type to type out an object shape that represents
// all ApplicationState keys and their string equivalents.
type ApplicationStateKeyTypes = {
  [K in ApplicationStateKeyType]: ApplicationStateKeyType
};
// Object that maps out all possible ApplicationState keys
export const ApplicationStateKey: ApplicationStateKeyTypes = {
  locationState: 'locationState',
  remoteDataState: 'remoteDataState',
  callState: 'callState',
  userStatsState: 'userStatsState',
  userState: 'userState'
};

const v4config = {
  whitelist: [
    ApplicationStateKey.locationState,
    ApplicationStateKey.userStatsState,
    ApplicationStateKey.userState,
    ApplicationStateKey.callState
  ]
};

const config = {
  key: 'fivecalls',
  debug: true,
  storage,
  whitelist: [
    ApplicationStateKey.locationState,
    ApplicationStateKey.userStatsState,
    ApplicationStateKey.userState,
    ApplicationStateKey.callState
  ],
  getStoredState: getStoredState(v4config)
};

const rootReducer = persistCombineReducers(config, {
  remoteDataState: remoteDataReducer,
  callState: callStateReducer,
  locationState: locationStateReducer,
  userStatsState: userStatsReducer,
  userState: userStateReducer
});

export default rootReducer;
