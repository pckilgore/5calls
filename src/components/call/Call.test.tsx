import * as React from 'react';
import { shallow } from 'enzyme';

import { Call } from './index';
import { Issue, ContactList } from '../../common/models';
import { CallState } from '../../redux/callState';
import { mockContact } from '../../common/models/contact';
import { mockIssue } from '../../common/models/issue';

test('Call component should be rendered if passed a valid object', () => {
  let callState: CallState = {
    currentIssueId: 'test1',
    contactIndexes: { test1: 2, test2: 1 },
    completedIssueIds: ['test1', 'test2']
  };
  let contactList = new ContactList();
  contactList.house = [mockContact];
  const component = shallow(
    <Call
      issue={mockIssue}
      contacts={contactList}
      callState={callState}
      getContactsIfNeeded={jest.fn()}
    />
  );
  expect(component).toMatchSnapshot();
});
