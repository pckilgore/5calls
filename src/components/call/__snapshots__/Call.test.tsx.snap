// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Call component should be rendered if passed a valid object 1`] = `
<section
  className="call"
>
  <Component
    currentIssue={
      Issue {
        "active": true,
        "contactAreas": Array [],
        "contactType": "REPS",
        "createdAt": "",
        "id": 123,
        "link": "",
        "linkTitle": "",
        "name": "Issue Name",
        "outcomeModels": Array [],
        "reason": "reason",
        "script": "issue *mock script*",
        "slug": "issue-name",
        "stats": Object {
          "calls": 10,
          "completion": 0,
        },
      }
    }
  />
  <ContextConsumer>
    <Component />
  </ContextConsumer>
</section>
`;
